pkg/cip/audit/auditor.go

```diff
@@ -0,0 +1,366 @@
+
+	if err := json.Unmarshal(body, &psm); err != nil {
+		return nil, fmt.Errorf("json.Unmarshal (request body): %v", err)
+	}
+
+	if err := json.Unmarshal(psm.Message.Data, &gcrPayload); err != nil {
+		return nil, fmt.Errorf("json.Unmarshal (message data): %v", err)
+	}
+
+	return &gcrPayload, nil
+}
+
+// ValidatePayload ensures that the payload is well-formed, per our
+// business-logic needs.
+func ValidatePayload(gcrPayload *reg.GCRPubSubPayload) error {
+	if gcrPayload.FQIN == "" && gcrPayload.PQIN == "" {
+		return fmt.Errorf(
+			"%v: neither 'digest' nor 'tag' was specified", gcrPayload)
+	}
+
+	if err := gcrPayload.PopulateExtraFields(); err != nil {
+		return err
+	}
+
+	switch gcrPayload.Action {
+	case "":
+		return fmt.Errorf("%v: Action not specified", gcrPayload)
+
+	// All deletions will for now be treated as an error. If it's an insertion,
+	// it can either have "digest" with FQIN, or "digest" + "tag" with PQIN. So
+	// we always verify FQIN, and if there is PQIN, verify that as well.
+	case "DELETE":
+		// Even though this is an error, we successfully processed this message,
+		// so exit with an error.
+		return fmt.Errorf("%v: deletions are prohibited", gcrPayload)
+	case "INSERT":
+		break
+	default:
+		return fmt.Errorf(
+			"%v: unknown action %q", gcrPayload, gcrPayload.Action)
+	}
+
+	return nil
+}
+
+// Audit receives and processes a Pub/Sub push message. It has 3 parts: (1)
+// parse the request body to understand the GCR state change, (2) update the Git
+// repo of the promoter manifests, and (3) reconcile these two against each
+// other.
+// nolint[funlen]
+func (s *ServerContext) Audit(w http.ResponseWriter, r *http.Request) {
+	logInfo := s.LoggingFacility.GetInfoLogger()
+	logError := s.LoggingFacility.GetErrorLogger()
+	logAlert := s.LoggingFacility.GetAlertLogger()
+
+	defer func() {
+		if msg := recover(); msg != nil {
+			panicStr := msg.(string)
+
+			stacktrace := debug.Stack()
+
+			s.ErrorReportingFacility.Report(
+				errorreporting.Entry{
+					Req:   r,
+					Error: fmt.Errorf("%s", panicStr),
+					Stack: stacktrace,
+				},
+			)
+
+			logAlert.Printf("%s\n%s\n", panicStr, string(stacktrace))
+			klog.Errorln(panicStr)
+		}
+	}()
+	// (1) Parse request payload.
+	gcrPayload, err := ParsePubSubMessage(r.Body)
+	if err != nil {
+		// It's important to fail any message we cannot parse, because this
+		// notifies us of any changes in how the messages are created in the
+		// first place.
+		msg := fmt.Sprintf("(%s) TRANSACTION REJECTED: parse failure: %v", s.ID, err)
+		_, _ = w.Write([]byte(msg))
+
+		panic(msg)
+	}
+
+	// Additionally, fail any message that we cannot validate. This is where we
+	// catch things like "DELETE" actions and warn them outright as all
+	// deletions are prohibited.
+	if err := ValidatePayload(gcrPayload); err != nil {
+		msg := fmt.Sprintf("(%s) TRANSACTION REJECTED: validation failure: %v", s.ID, err)
+		_, _ = w.Write([]byte(msg))
+
+		panic(msg)
```

Comment:
there is a reason we panic here and in places? and in other parts, we log the error and return.
more a clarify question to me :)

Reply:
> p.s. this is not new code; just a copy of container-image-promoter + lint fixes, so no need to review for structure.
(there will be a refactor to follow.)

@cpanato -- :point_up: FWIW, we should probably be returning errors instead of panicking, but this was meant to be a 1:1 copy, as opposed to a refactor. We'll still need to dig in to better understand everything.

Reply:
thanks!!